### PR TITLE
Fix case where tracking batch is never updated in batched full join

### PR DESCRIPTION
**Pull request by jbrennan333**
Signed-off-by: Jim Brennan <jimb@nvidia.com>

Closes https://github.com/NVIDIA/spark-rapids/issues/7516.

As reported in https://github.com/NVIDIA/spark-rapids/issues/7516, some of our nightly build tests running on the yarn cluster were reporting failures for some full join test cases.

From examining the CPU vs GPU output for some of the tests, it appeared that the final batch (which represents rows on the build side that were not matched on the streaming side) was empty where it shouldn't be.  Once I was able to reproduce on the yarn cluster, I confirmed this with debug logging.

In cases where there was nothing on the stream side for a task, we were never updating the tracking mask to indicate which rows from the build side were needed for the final batch.

With this fix, the tests on the yarn cluster now pass, and all full join integration tests also pass for me locally.

### CODE REVIEW

**Comment by jbrennan333** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala [hunk 1]
```diff
@@ -601,6 +601,14 @@ class HashFullJoinIterator(
       try {
         val lazyLeftMap = LazySpillableGatherMap(maps(0), spillCallback, "left_map")
         val lazyRightMap = LazySpillableGatherMap(maps(1), spillCallback, "right_map")
+        withResource(new NvtxWithMetrics("update tracking mask",
+          NvtxColor.ORANGE, joinTime)) { _ =>
+          closeOnExcept(lazyLeftMap) { lazyLeft =>
+            closeOnExcept(lazyRightMap) { lazyRight =>
+              updateTrackingMask(if (buildSide == GpuBuildRight) lazyRight else lazyLeft)
+            }
+          }
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:610]: nit: could be slightly more compact
```suggestion
          closeOnExcept(Seq(lazyLeftMap, lazyRightMap)) { _ =>
            updateTrackingMask(if (buildSide == GpuBuildRight) lazyRightMap else lazyLeftMap)
          }
```

**Review by gerashegalov — DISMISSED** [review]
LGTM overall

would be good to have a test reproducing it without YARN

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala (hunk 1 above)
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala:610]: Thanks for the suggestion.  I will update.

**Comment by jbrennan333** [comment]
@gerashegalov thanks for the review.  I am going to attempt to create an integration test for this, but I'm not certain I'll be able to - my attempts to repro locally so far have not worked.   If I can get one to work, I would like to put it up as a follow-on PR instead of holding up this one.  I am planning to update this to address the nit you mentioned above.

**Comment by revans2** [comment]
@jbrennan333 it should be fairly simple to create an integration test for this. We would need/want each side of the join to have a very low cardinality (smaller than the number of shuffle partitions), and be disjoined. We can also explicitly set the number of shuffle partitions to be a high number, and change the values in lhs and rhs if needed so we know that they don't end up hashing to the same thing.

```
lhs = spark.createDataframe([1] * 100, ...)
rhs = spark.createDataframe([2] * 100, ...)
// Do the join...
```

We also would want to make sure that we don't accidentally do a broadcast join.  This should let us have at least one task with a build side but no stream side data. Also a stream side with not build side and at least a few with nothing in each.

Might be good to include this same test for all of the join types.

**Comment by jbrennan333** [comment]
Thanks @revans2.  I was actually able to reproduce with `test_broadcast_join_left_table` by just setting `spark.sql.shuffle.partitions=200`, which I think was the key conf that made it happen on yarn.  I'm double checking that the fix works with this change.

**Review by revans2 — APPROVED** [review]
Would like to see the test. I was able to make it happen in scala.

```
val lhs = List.fill(100)(1).toDF
val rhs = List.fill(100)(2).toDF
lhs.join(rhs, lhs("value") === rhs("value"), joinType="fullouter").groupBy(lhs("value")).agg(count(lit(1))).show()
+-----+--------+
|value|count(1)|
+-----+--------+
|    1|     100|
+-----+--------+

spark.conf.set("spark.rapids.sql.enabled", "false")
lhs.join(rhs, lhs("value") === rhs("value"), joinType="fullouter").groupBy(lhs("value")).agg(count(lit(1))).show()
+-----+--------+
|value|count(1)|
+-----+--------+
| null|     100|
|    1|     100|
+-----+--------+
```

**Comment by jbrennan333** [comment]
I just modified this test to add the conf:
```
def test_broadcast_join_left_table(data_gen, join_type):
    def do_join(spark):
        left, right = create_df(spark, data_gen, 250, 500)
        return broadcast(left).join(right, left.a == right.r_a, join_type)
    assert_gpu_and_cpu_are_equal_collect(do_join, conf={'spark.sql.shuffle.partitions': '200'})
```

**Comment by jbrennan333** [comment]
I pushed up the change above. 
@revans2, let me know if you think I should split it out to a separate test, or do something more explicit.

File: integration_tests/src/main/python/join_test.py [hunk 2]
```diff
@@ -487,7 +487,7 @@ def test_broadcast_join_left_table(data_gen, join_type):
     def do_join(spark):
```
Comment (jlowe) [inline integration_tests/src/main/python/join_test.py:487]: 2023 copyrights

File: integration_tests/src/main/python/join_test.py [hunk 3]
```diff
@@ -487,7 +487,7 @@ def test_broadcast_join_left_table(data_gen, join_type):
...
         left, right = create_df(spark, data_gen, 250, 500)
         return broadcast(left).join(right, left.a == right.r_a, join_type)
-    assert_gpu_and_cpu_are_equal_collect(do_join)
+    assert_gpu_and_cpu_are_equal_collect(do_join, conf={'spark.sql.shuffle.partitions': '200'})
```
Comment (jlowe) [inline integration_tests/src/main/python/join_test.py:490]: There should be a comment here explaining why this has a custom number or partitions but other tests do not.

File: integration_tests/src/main/python/join_test.py (hunk 2 above)
Comment (jbrennan333) [reply integration_tests/src/main/python/join_test.py:487]: Thanks.

File: integration_tests/src/main/python/join_test.py (hunk 3 above)
Comment (jbrennan333) [reply integration_tests/src/main/python/join_test.py:492]: I added a comment.

**Comment by jbrennan333** [comment]
build

File: integration_tests/src/main/python/join_test.py (hunk 3 above)
Comment (gerashegalov) [reply integration_tests/src/main/python/join_test.py:492]: aren't we losing test coverage for the previously effective default value of shuffle.partitions. It feels safer If we parametrize the test with the empty conf `{}` and `{'spark.sql.shuffle.partitions': 200}`

Comment (jbrennan333) [reply integration_tests/src/main/python/join_test.py]: I made this change.

File: integration_tests/src/main/python/join_test.py [hunk 4]
```diff
@@ -487,7 +487,9 @@ def test_broadcast_join_left_table(data_gen, join_type):
     def do_join(spark):
         left, right = create_df(spark, data_gen, 250, 500)
         return broadcast(left).join(right, left.a == right.r_a, join_type)
-    assert_gpu_and_cpu_are_equal_collect(do_join)
+    # Specify 200 shuffle partitions to test cases where streaming side is empty
+    # as in https://github.com/NVIDIA/spark-rapids/issues/7516
+    assert_gpu_and_cpu_are_equal_collect(do_join, conf={'spark.sql.shuffle.partitions': '200'})
```
[the remark this answers is not in the document]
Comment (jbrennan333) [reply integration_tests/src/main/python/join_test.py]: fixed in the parametrized version.

**Comment by jbrennan333** [comment]
build

**Review by gerashegalov — APPROVED** [review]

**Review by jlowe — APPROVED** [review]

**Pull request merged**